Build: Use ESLint instead of JSHint/JSCS

platinumazure opened the pull request:
Work in progress: Just fixing minor issues at this point.

TODOs:

* [x] Fix indentation issues
* [x] Use `eslint-disable-next-line` comments instead of `eslint-disable-line` comments

platinumazure commented:
Cripes, this will fail on Node <4.x due to ESLint no longer supporting that environment. Should I move the ESLint Grunt target into the build section so it's run on latest Node?

leobalter commented:
> Should I move the ESLint Grunt target into the build section so it's run on latest Node?

Yes. we can run ESLint only in a single version of Node within Travis.

There's a package.json script called 'ci' calling grunt. We should redefine those to exclude linting tasks and make a special call for it on travis latest Node.

platinumazure commented:
Okay, I've pushed a commit which fixes the ESLint issues in tests.

test/main/deepEqual.js proved particularly problematic, probably because it is testing some extreme cases. So here is what I did in that file:
* Fixed space-before-function-paren, space-in-parens, and one semi-spacing issue
* Replaced a `/* jshint asi:true */` with equivalent `/* eslint-disable-next-line: semi */`
* Disabled `max-len` (no more than 100 characters/line), `array-bracket-spacing` (require space just inside array brackets), and `comma-spacing` (require space after commas) for that whole file.

If you want me to do something different for that file (e.g., fix the `comma-spacing` issues), let me know and I can push another commit.

Everything else was pretty straightforward, but let me know if I missed anything important.

platinumazure commented:
@leobalter I could use some help configuring Travis as you suggested. Also, can you confirm that AppVeyor intentionally only runs on latest Node and so no reconfiguration there will be necessary? Thanks!

trentmwillis commented:
> What is the preferred IIFE-wrapping style?

Looks like both have examples in our JS style guide. I like the [AirBnb approach](https://github.com/airbnb/javascript#functions--iife) to this. Additionally, I have heard, that using the `(function() {}())` form is better optimized in V8 (but that shouldn't matter much to us).

platinumazure commented:
Okay- their style guide says either way is fine, though their ESLint
configuration prefers wrapping the CallExpression. Wrapping the
CallExpression is also the mode that you said might be optimized better in
v8.

Here's what I propose. Since there is inconsistency right now, I will
ensure IIFEs are wrapped but allow both styles. Then in a future issue, if
someone wants to make the case for bringing consistency into the codebase
by switching to CallExpression/"outer" wrapping everywhere, we can evaluate
then.

@trentmwillis How does that sound?

On Aug 23, 2016 10:14 PM, "Trent Willis" <notifications@github.com> wrote:

> What is the preferred IIFE-wrapping style?
>
> Looks like both have examples in our JS style guide. I like the AirBnb
> approach <https://github.com/airbnb/javascript#functions--iife> to this.
> Additionally, I have heard, that using the (function() {}()) form is
> better optimized in V8 (but that shouldn't matter much to us).
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jquery/qunit/pull/1028#issuecomment-241946153>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AARWeqglcWZY4PLH0XbapMoYAh7q43_9ks5qi7cVgaJpZM4JfuM0>
> .
>

mgol commented:
It sounds surprising that one IIFE form may be better optimized (whatever
that means) in v8 - it all disappears anyway when the code is JIT'ted.
Also, you'd have to create IIFEs in loops to see any difference, if it's
just for wrapping a few files then it shouldn't matter in the slightest.

On Wednesday, 24 August 2016, Kevin Partington <notifications@github.com>
wrote:

> Okay- their style guide says either way is fine, though their ESLint
> configuration prefers wrapping the CallExpression. Wrapping the
> CallExpression is also the mode that you said might be optimized better in
> v8.
>
> Here's what I propose. Since there is inconsistency right now, I will
> ensure IIFEs are wrapped but allow both styles. Then in a future issue, if
> someone wants to make the case for bringing consistency into the codebase
> by switching to CallExpression/"outer" wrapping everywhere, we can evaluate
> then.
>
> @trentmwillis How does that sound?
>
> On Aug 23, 2016 10:14 PM, "Trent Willis" <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:
>
> > What is the preferred IIFE-wrapping style?
> >
> > Looks like both have examples in our JS style guide. I like the AirBnb
> > approach <https://github.com/airbnb/javascript#functions--iife> to this.
> > Additionally, I have heard, that using the (function() {}()) form is
> > better optimized in V8 (but that shouldn't matter much to us).
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/jquery/qunit/pull/1028#issuecomment-241946153>, or
> mute
> > the thread
> > <https://github.com/notifications/unsubscribe-auth/
> AARWeqglcWZY4PLH0XbapMoYAh7q43_9ks5qi7cVgaJpZM4JfuM0>
> > .
> >
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jquery/qunit/pull/1028#issuecomment-241947186>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABrUnu6W6f1mpPsRWGL4XPmYdZWe0bX-ks5qi7k8gaJpZM4JfuM0>
> .
>


-- 
Michał Gołębiowski

platinumazure commented:
@mgol Apologies, I thought I had responded to your comment. I'm going to interpret your comment as "IIFE wrapping style shouldn't matter" and thus continue with [my current proposal](https://github.com/jquery/qunit/pull/1028#issuecomment-241947186). Let me know if you object.

platinumazure commented:
@leobalter Okay, I think I've figured things out. I just pushed a commit which adds a package.json script "ci-include-lint", which runs `grunt eslint && grunt test && grunt coveralls`. In .travis.yml, I've added an extra build node to the matrix which runs that script. An example of how that job would look can be found [here](https://travis-ci.org/jquery/qunit/jobs/157520687).

I think this gets us 95% of the way there. I think that maybe instead of "ci-include-lint" (which also runs the same CI job that is run on the default Node 6 build node), I should create a "lint" script which would only run ESLint on Node 6, that way we don't run more CI jobs than we need to. If you see this before I get a chance to try that experiment, please do let me know if you think that's a good direction.

leobalter commented:
> I should create a "lint" script which would only run ESLint on Node 6, that way we don't run more CI jobs than we need to. 

That's the first thing I noticed. Also the current way repeats the other tasks on ci-include-lint.

This is looking good, almost ready to go! ++

platinumazure commented:
@leobalter Not sure I fully understood your reply. Do you want me to avoid creating an extra build node altogether? Or can I create that node (as I am currently doing), as long as it only lints rather than running the CI scripts as well?

I was proposing the latter, but if you would prefer I find another way, please tell me what you have in mind.

mgol commented:
@platinumazure Currently you're invoking both `ci` and `ci-include-lint` on Node 6 which duplicates most stuff. Since Node 6 already runs the `ci-include-lint` task it shouldn't run the `ci` one. That's, I think, what @leobalter says.

platinumazure commented:
Oh. I was going to go the other way and create a "lint" task (and remove
"ci-include-lint") which the extra node would run, plus have a Node 6 box
that still runs CI. Those seem equivalent to me except for one more build
node (and therefore one more npm install at least) in my proposal.

If you can advise me on how I can conditionally remove/set the *existing*
Node 6 build node to use ci-include-lint *instead of* ci in Travis config,
then I'm certainly okay with doing that too. This is the first time I've
had to go build matrix configuration with Travis, so I'm definitely taking
shots in the dark here.

I won't be able to make any modifications until tonight (US time) at
earliest anyway, so no huge rush. Thanks in advance!

On Sep 5, 2016 7:15 AM, "Michał Gołębiowski" <notifications@github.com>
wrote:

> @platinumazure <https://github.com/platinumazure> Currently you're
> invoking both ci and ci-include-lint on Node 6 which duplicates most
> stuff. Since Node 6 already runs the ci-include-lint task it shouldn't
> run the ci one. That's, I think, what @leobalter
> <https://github.com/leobalter> says.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jquery/qunit/pull/1028#issuecomment-244733047>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AARWeoQkwwUjh3yfmBx_tnzirc0xmOmFks5qnAfbgaJpZM4JfuM0>
> .
>

mgol commented:
> If you can advise me on how I can conditionally remove/set the *existing*
> Node 6 build node to use ci-include-lint *instead of* ci in Travis config,
> then I'm certainly okay with doing that too.

The following URL should explain it a little:
https://docs.travis-ci.com/user/customizing-the-build#Build-Matrix

platinumazure commented:
@mgol Thanks for the link! I already skimmed that to get as far as I have, but I guess I had to dig deeper. :smile:

I *think* this latest commit should be pretty much good enough to merge, unless I've missed something. By all means, please review and let me know if I should do something different (either in Travis config or elsewhere). I can squash if necessary, as well (just leaving commit history visible for now).

@leobalter Think I've figured out the Travis config now, please review at your convenience. Latest PR build can be found here: https://travis-ci.org/jquery/qunit/builds/157738305

leobalter commented on Gruntfile.js:
```diff
@@ -37,36 +37,28 @@ grunt.initConfig( {
+    eslint: {
+        options: {
+            config: ".eslintrc.json"
+        },
```
We've got some weird indentation in this file.

platinumazure replied:
Dagnabbit, I'll fix ASAP.

platinumazure commented:
Oh, and I could still use some feedback on my second big question: In tests/main/deepEqual.js, does the balance of rules enabled (with violations fixed) vs rules selectively disabled/ignored feel okay? Thanks!

leobalter commented on src/dump.js:
```diff
@@ -110,7 +110,11 @@ export default ( function() {
+                    return this.HTML ? "<br />" : "\n";
+                } else {
+                    return this.HTML ? "&#160;" : " ";
+                }
```
indentation again

leobalter commented on src/test.js:
```diff
@@ -374,7 +374,9 @@ Test.prototype = {
-					function() { resume(); },
+					function() {
+						resume();
+					},
```
linting hurts our feelings

platinumazure replied:
I'll look into this-- that might have just been me being OCD. (But I'm not 100% sure.)

leobalter replied:
You don't need to, that's also ok. There isn't any real issue here.

platinumazure replied:
Turns out it was me after all (I thought there was a violation earlier but I must have tweaked a config to make that okay). I'll revert this.

leobalter commented on src/test.js:
```diff
@@ -719,15 +727,15 @@ function internalStart( test ) {
 function numberOfTests( module ) {
 	var count = module.tests.length;
-	while ( module = module.childModule ) {
+	while ( ( module = module.childModule ) ) {
```
why?

platinumazure replied:
Yeah, this is one I probably should have called out in my initial writeup.

ESLint has a rule `no-cond-assign`, which flags against assignments in conditional parts of a control statement (the assumption being the developer probably meant to use `==` or `===`). That rule can be configured to ignore such assignments if they are in redundant parentheses. In other words, a developer can signal his/her intention to assign in a condition by adding parentheses. (It looks a lot more reasonable without spaces inside parentheses, though.)

So we have three options here:
* Have ESLint ignore the violation on this line, but leave the rule on and don't use extra parentheses
* Use extra parentheses (this is what is happening now in my PR)
* Disable the rule, if we trust that developers know what they are doing in this space

How would you like me to proceed?

platinumazure replied:
Forgot to add: [Here's the section](https://github.com/jquery/eslint-config-jquery/blob/master/.eslintrc.json#L4-L7) in eslint-config-jquery showing the "allow in parentheses" configuration I was describing. In other words, I didn't turn on the rule personally, but rather it came from the config we have chosen to use.

Again, if we want to change how it works in this repo, that's possible and totally okay.

leobalter replied:
Thanks for the clarification. I'm fine with it as it might prevent unintended uses and the extra parenthesis flags it as intentional.

platinumazure replied:
Awesome, I will make no change here then. :+1:

leobalter commented:
> In tests/main/deepEqual.js, does the balance of rules enabled (with violations fixed) vs rules selectively disabled/ignored feel okay? 

Well, that file is a nightmare for linting tools. It feels fine for me in any case as I've passed some long moments trying to get it better and I know how hard it is to keep it ok. I would say let's keep the current way it is and see if we get any feedback in the future.

platinumazure commented:
Okay, I *think* all open questions have been addressed at this point. Let me know if I should squash. Thanks!

leobalter commented on src/assert.js:
```diff
@@ -119,19 +119,17 @@ Assert.prototype = {
-		/*jshint eqeqeq:false */
 		this.pushResult( {
-			result: expected == actual,
+			result: expected == actual,		// eslint-disable-line eqeqeq
```
can we use the eslint-disable-next-line form? this repeats in line 132

platinumazure replied:
We can definitely do that, but there's another rule we would need to make a decision about.

Relevant configuration from eslint-config-jquery:
> 		"lines-around-comment": [
> 			"error",
> 			{
> 				"beforeLineComment": true
> 			}
> 		],

Basically, I would need to add a blank line before the comment. That could be okay, but I noticed the rest of the code was pretty condensed and didn't want to be adding a bunch of line breaks in the middle of objects without being sure it was okay. Using the inline form was more "convenient" in that regard because I could leave the existing structure.

How should I proceed?

* No changes (leave "eslint-disable-line" comments)
* Switch to eslint-disable-next-line and add necessary line breaks to conform with lines-around-comment
* Switch to eslint-disable-next-line and turn off the lines-around-comments rule in QUnit's ESLint configuration

leobalter replied:
We historically use the comment in the line above after a blank line. So I'm biased towards that as I am used to it. There are other in-comments rules placed like this, so I'm again leaning towards this way.

The second option looks like the best, but I'm open to discuss.

platinumazure replied:
No problem, I'm okay with the second option! I just wanted to be sure that's what you wanted. I'll implement that hopefully no later than tomorrow evening (US time).

leobalter commented:
LGTM, we've only got some remaining whitespace issues but those can be fixed during merge.

The Gruntfile.js file still has some indentation issues, and there are these [inline](https://github.com/jquery/qunit/pull/1028#discussion-diff-77624731) rules which I would use the comment above form. Do you prefer sticking with the inline style?

platinumazure commented:
@leobalter Left a response to your comment on eslint-disable-line.

Also, regarding indentation: I can try to find remaining indentation issues by using the ESLint "indent" rule. It's not enabled in eslint-config-jquery because the rule does not elegantly handle the top-level IIFE requirement in the jQuery style guide, but I can at least go through and try to fix any real errors. Sound okay?

leobalter commented:
On the indentation, I'm ok with that but the current issues are not many so we can merge this PR before and open a follow up. As you prefer.

platinumazure commented:
I found a few files that had two-space indentation (these are files I
haven't yet touched). Shall I fix those up as well?

On Sep 6, 2016 8:23 AM, "Leo Balter" <notifications@github.com> wrote:

> On the indentation, I'm ok with that but the current issues are not many
> so we can merge this PR before and open a follow up. As you prefer.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jquery/qunit/pull/1028#issuecomment-244948125>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AARWeu0zdbZosnoqviMs1bv7X7E30K8_ks5qnWkugaJpZM4JfuM0>
> .
>

platinumazure commented:
@leobalter I've pushed a commit replacing `eslint-disable-line` with `eslint-disable-next-line`.

Regarding the indentation issues: I've been looking and I see a number of indentation problems throughout the repository (about 200 or so violations), many in files I didn't touch. So I guess I'd like to treat those as a separate issue. I am happy to fix any indentation problems I've introduced in this PR, if you can help me find them. Otherwise, can we merge? (Assuming there aren't other issues, of course.)

Let me know if you need me to rebase and/or squash.

platinumazure commented:
Pinging @leobalter. Can you please let me know which indent errors you want me to fix?

Also, I'm happy to report that a long-standing bug with the `indent` rule has been fixed and will be released with ESLint 3.6.0: If tabs are required but spaces are present, it would report "0 tabs" instead of "n spaces", or vice versa. So we could take a serious look at indentation issues at that time, if desired.

leobalter commented:
I'm sorry I didn't answer this before, let get this done and merged

platinumazure commented:
@leobalter No worries. As far as I'm concerned this is ready except for "some remaining whitespace issues". I fixed the Gruntfile.js indentation and any other mistakes I had introduced, as far as I know. Are there any other indentation problems that I introduced that I should fix, or can we merge?

leobalter commented:
I was already merging this, just runned a final batch of tests to confirm.

I've got it squashed on 89731f9.

Thanks, that's a great work!

platinumazure commented:
Thanks! Sorry for pestering you when you were merging, I wasn't sure if you were still waiting on me.

leobalter commented:
No worries at all, you can ping anytime!
